### PR TITLE
ast/compile: unused imports follow up

### PR DESCRIPTION
**Pull request by srenatus**
This is a folllow-up to #4377.

The approach I had advised to take in #4377 fell a little short in the case of shadowed vars: Considering

```rego
package p
import data.foo # unused
r { f(1) }
f(foo) = foo == 1
```
and
```rego
package p
import data.foo # unused
r { foo := true; foo }
```

we have a really hard time figuring out that the vars in those bodies are _not_ actually making use of the imports.

To remedy this, we'll need to do the same work done in resolveAllRefs, which turned out to be much more involved than an `ast.Walk*`.

In the last commit, we're merging the check with the resolveAllRefs stage, because it's avoiding some extra work; and gives us better coverage if there are ever any lazily loaded modules.

### CODE REVIEW

File: ast/compile.go [hunk 1]
```diff
@@ -1468,6 +1411,21 @@ func (c *Compiler) resolveAllRefs() {
 			return false
 		})
 
+		if c.strict { // check for unused imports
```
Comment (johanfylling) [inline ast/compile.go:1414]: Consider updating the docs of this function to reflect this added responsibility.

File: ast/compile_test.go [hunk 2]
```diff
@@ -1594,48 +1587,85 @@ func TestCompilerCheckUnusedImports(t *testing.T) {
...
+
+			r { foo := true; foo }
+			`,
+			expectedErrors: Errors{
+				&Error{
+					Location: NewLocation([]byte("import"), "", 2, 4),
+					Message:  "import data.foo unused",
+				},
+			},
+		},
```
Comment (johanfylling) [inline ast/compile_test.go:1668]: Should we also add test cases for when an import is used/not used as a rule value? E.g.
```rego
package p
import data.foo # unused
import data.bar

r = bar { true }
```
and
```rego
package p
import data.foo # unused
import data.bar #unused

r = data.bar { true }
```

**Review by johanfylling — APPROVED** [review]
LGTM!

File: ast/compile.go (hunk 1 above)
Comment (srenatus) [reply ast/compile.go:1414]: That has been added with the PR that this is a follow-up of, I think.

**Comment by srenatus** [comment]
Will merge when green ✅

**Pull request merged**